A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+child policy as described below.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing and just pass the RPC through (to the next
+filter). For example if the `path` configuration value is `/Package1.Service2`
+and the RPC method is `/Package2.Service1/Method3` then just pass the RPC
+through.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
+warning and pass the RPC through. For a valid `override-host` value set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path of the RPC, if the RPC path does not match the
+  `path` config value of the filter then skip further processing and
+  pass the response through. Otherwise check if `upstreamAddress` is
+  set (in the RPC context of this filter) and get the value. Also get
+  the peer address for the RPC; let’s call it `hostAddress` here. For
+  example, in Java you get the peer address through the
+  `Grpc.TRANSPORT_ATTR_REMOTE_ADDR` attribute of the current
+  ClientCall. If `upstreamAddress` is not set or `upstreamAddress` and
```

Comment:
I am not sure this is entirely consistent with Envoy's implementation of the same functionality. I think Envoy will return a `Set-Cookie` header on every response.

Reply:
My understanding is based on the code [here](https://github.com/envoyproxy/envoy/blob/56a8223a45608dd066f854486b81c9f62bee6ebe/source/extensions/http/stateful_session/cookie/cookie.cc#L11),  which is :

```C++
  if (!upstream_address_.has_value() || host_address != upstream_address_.value()) {
    const std::string encoded_address =
        Envoy::Base64::encode(host_address.data(), host_address.length());
    headers.addReferenceKey(Envoy::Http::Headers::get().SetCookie,
                            factory_.makeSetCookie(encoded_address));
  }
```

on which I based the pseudo-code snippet here. But it's easy for us to always set `Set-Cookie` if that's the desired behavior (or my understanding of the Envoy code is wrong).

Reply:
I think I remember seeing that Envoy always sends back the "Set-Cookie" header but I would need to confirm.

Reply:
Okay, I will wait for your findings

Reply:
okay, confirmed that Envoy sends the Set-Cookie header only on the first request (when Cookie is not provided by the client).

Reply:
Thanks for confirming.